File: problem-set-2/readability/readability.c
#include <cs50.h>
#include <ctype.h>
#include <math.h>
#include <stdio.h>
#include <string.h>

int calculate_number_of_letters(string text);
int calculate_number_of_words(string text);
int calculate_number_of_sentences(string text);
float calculate_reading_level(int letterCount, int wordCount, int sentenceCount);
void display_reading_level(float readingLevel);

int main(void)
{
    // Prompt user for a text input
    string text = get_string("Text: ");

    int letterCount = calculate_number_of_letters(text);
    int wordCount = calculate_number_of_words(text);
    int sentenceCount = calculate_number_of_sentences(text);

    float readingLevel = calculate_reading_level(letterCount, wordCount, sentenceCount);

    display_reading_level(readingLevel);
}

// Declare a function to get the total number of letters in the text.
int calculate_number_of_letters(string text)
{
    int letterCount = 0;

    for (int index = 0; text[index] != '\0'; index++)
    {
        if (isalpha(text[index]))
        {
            letterCount++;
        }
    }

    return letterCount;
}

// Declare a function to get the total number of words in the text
int calculate_number_of_words(string text)
{
    int wordCount = 0;

    for (int index = 0; text[index] != '\0'; index++)
    {
        if (isspace(text[index]))
        {
            wordCount++;
        }
    }

    // Include the last word in the text, which will be skipped by the loop
    wordCount++;

    return wordCount;
}

// Declare a function to get the total number of sentences in the text
int calculate_number_of_sentences(string text)
{
    int sentenceCount = 0;

    for (int index = 0; text[index] != '\0'; index++)
    {
        if (text[index] == '.' || text[index] == '?' || text[index] == '!')
        {
            sentenceCount++;
        }
    }

    return sentenceCount;
}

// Declare a function to get the reading level of a text
float calculate_reading_level(int letterCount, int wordCount, int sentenceCount)
{
    float L = ((float) letterCount / (float) wordCount) * 100.0;
    float S = ((float) sentenceCount / (float) wordCount) * 100.0;

    float readingLevel = 0.0588 * L - 0.296 * S - 15.8;

    return readingLevel;
}

void display_reading_level(float readingLevel)
{
    int grade = rintf(readingLevel);

    if (grade >= 16)
    {
        printf("Grade 16+\n");
    }
    else if (grade < 1)
    {
        printf("Before Grade 1\n");
    }
    else
    {
        printf("Grade %i\n", grade);
    }
}

File: problem-set-6/sentimental-readability/readability.py
# Checks the reading level of a piece of text using the Coleman-Liau index

from cs50 import get_string


def main():
    # Get text input from user
    text = get_string("Text: ")

    # Initialize counters
    letter_count = 0
    word_count = 0
    sentence_count = 0

    # Iterate over the text input to count the number of letters, words, and sentences
    for character in text:

        # Count all alphabetic letters
        if character.isalpha():
            letter_count += 1

        # Count all words
        elif character.isspace():
            word_count += 1

        # Count all sentences
        elif character in ['.', '?', '!']:
            sentence_count += 1

    # Add one to word_count to account for the last word in the sentence
    word_count += 1

    # Calculate the reading level
    reading_level = calculate_reading_level(letter_count, word_count, sentence_count)

    # Print the reading level
    print_reading_level(reading_level)


# Calculates reading level using the Coleman-Liau index
def calculate_reading_level(letter_count, word_count, sentence_count):
    # Calculate the average number of letters per 100 words
    L = 100 * (letter_count / word_count)

    # Calculate the average number of sentences per 100 words
    S = 100 * (sentence_count / word_count)

    # Use Coleman-Liau index to compute the reading level of the text input
    return round(0.0588 * L - 0.296 * S - 15.8)


# Prints reading level of text input
def print_reading_level(reading_level):
    if reading_level < 1:
        print(f"Before Grade 1")
    elif reading_level >= 16:
        print(f"Grade 16+")
    else:
        print(f"Grade {reading_level}")


main()
